Elasticsearch: Add interval type selector to interval field

matyax opened the pull request:
This PR is a follow up to the [recent addition](https://github.com/grafana/grafana/pull/75459) to support calendar_interval in histograms, with the objective of making it easier for users to:
- Understand that calendar_interval is supported.
- Improve the form to display the appropriate suggestions for each interval mode.

The way this was implemented is backwards compatible, doesn't change data structures, and it's purely visual.
 
**Which issue(s) does this PR fix?**:

Part of https://github.com/grafana/grafana/discussions/55034

**Special notes for your reviewer:**

See demo:

https://github.com/grafana/grafana/assets/1069378/1f0c38e6-35d9-4c67-9a8f-1bfff6e89974

matyax committed:
Elasticsearch: add interval type selector

matyax committed:
Interval type: add tooltip

matyax committed:
DateHistogramSettingsEditor: create unit test

ivanahuckova reviewed (commented):
This is slightly strange cause you can change interval type, but it only changes the options in dropdown. And if you select with `fixed interval` option to `1y`, it is going to behave like `calendar interval`. Have you considered just adding a tooltip/explanation when user selects interval that will be queried as `calendar interval`. Or if we want a interval type, then we should store it and use it accordingly. 

Also why are some of calendar_interval options available and others are not https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-datehistogram-aggregation.html?

matyax commented:
That is a good point, that I also realized while showing the changes today. I'll think about it and update the implementation. Thanks!

matyax commented on public/app/plugins/datasource/elasticsearch/QueryBuilder.ts:
```diff
@@ -26,6 +26,8 @@ import {
 } from './types';
 import { convertOrderByToMetricId, getScriptValue } from './utils';
 
+export const calendarIntervals: string[] = ['1w', '1M', '1q', '1y'];
```
> if we could <missing> this and just use

I might imagine what you mean, but I want to be sure. What was the actual suggestion?

ivanahuckova replied:
I would also maybe add a comment that there are other calendar intervals (`1m`,`1d`) but for now we have decided to support only those larger than `1d` as that seems to be the main usecase and doesn't need a lot of changes.

matyax replied:
Comment added.

matyax committed:
Elastic histogram settings: refactor showing calendar type

matyax committed:
Update test

matyax commented:
@ivanahuckova Updated! Removed the type selector, as it didn't make a lot of sense in this version, and changed it to update the label according to the selected option, which are now merged (calendar and fixed together).

Again, this is meant to be an improvement to the recent contribution, to make it more clear to the user how it's going to be queried.

> Also why are some of calendar_interval options available and others are not 

Basically because the contribution focused on calendar intervals larger than 1 day, which also seemed to be the most mentioned in the [feature request](https://github.com/grafana/grafana/discussions/55034) (1 week, 1 month, 1 quarter, etc), as are more variable than those <= 1 day.

This is something that we can definitely improve in the future, if needed.

matyax committed:
Prettier

ivanahuckova reviewed (approved):
Nice! Left some minor nits/comments but overall looks good.

matyax commented on public/app/plugins/datasource/elasticsearch/components/QueryEditor/BucketAggregationsEditor/SettingsEditor/DateHistogramSettingsEditor.tsx:
```diff
@@ -48,16 +53,33 @@ interface Props {
+      <InlineField
+        label={intervalType}
+        tooltip={
+          'Calendar-aware intervals adapt to varying day lengths, month durations, and leap seconds, considering the calendar context. Fixed intervals remain constant, always being multiples of SI units, independent of calendar changes.'
```
Absolutely, let's do that.

matyax committed:
DateHistogramSettingsEditor: change tooltip according to interval type

matyax committed:
Calendar intervals: add comment

matyax committed:
Prettier

matyax committed:
useCreatableSelectPersistedBehaviour: revert change
